flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobExecutionITCase.java

```diff
@@ -0,0 +1,144 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.jobmaster;
+
+import org.apache.flink.runtime.execution.Environment;
+import org.apache.flink.runtime.io.network.api.reader.RecordReader;
+import org.apache.flink.runtime.io.network.api.writer.RecordWriter;
+import org.apache.flink.runtime.io.network.partition.ResultPartitionType;
+import org.apache.flink.runtime.jobgraph.DistributionPattern;
+import org.apache.flink.runtime.jobgraph.JobGraph;
+import org.apache.flink.runtime.jobgraph.JobVertex;
+import org.apache.flink.runtime.jobgraph.tasks.AbstractInvokable;
+import org.apache.flink.runtime.jobmanager.scheduler.SlotSharingGroup;
+import org.apache.flink.runtime.minicluster.MiniCluster;
+import org.apache.flink.runtime.testutils.MiniClusterResource;
+import org.apache.flink.runtime.testutils.MiniClusterResourceConfiguration;
+import org.apache.flink.types.IntValue;
+import org.apache.flink.util.TestLogger;
+
+import org.junit.ClassRule;
+import org.junit.Test;
+
+import java.util.concurrent.CompletableFuture;
+
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assert.assertThat;
+
+/**
+ * Integration tests for job scheduling.
+ */
+public class JobExecutionITCase extends TestLogger {
+
+	private static final int SLOTS_PER_TM = 11;
+	private static final int NUM_TMS = 2;
+	private static final int PARALLELISM = SLOTS_PER_TM * NUM_TMS;
+
+	@ClassRule
+	public static final MiniClusterResource MINI_CLUSTER_RESOURCE = new MiniClusterResource(
+		new MiniClusterResourceConfiguration.Builder()
+			.setNumberSlotsPerTaskManager(SLOTS_PER_TM)
+			.setNumberTaskManagers(NUM_TMS)
+			.build());
+
+	@Test
+	public void testCoLocationConstraintJobExecution() throws Exception {
+		final JobGraph jobGraph = createJobGraph(PARALLELISM);
+
+		final MiniCluster miniCluster = MINI_CLUSTER_RESOURCE.getMiniCluster();
+
+		miniCluster.submitJob(jobGraph).get();
+
+		final CompletableFuture<JobResult> jobResultFuture = miniCluster.requestJobResult(jobGraph.getJobID());
+
+		assertThat(jobResultFuture.get().isSuccess(), is(true));
+	}
+
+	private JobGraph createJobGraph(int parallelism) {
+		final JobVertex sender = new JobVertex("Sender");
+		sender.setParallelism(parallelism);
+		sender.setInvokableClass(Sender.class);
+
+		final JobVertex receiver = new JobVertex("Receiver");
+		receiver.setParallelism(parallelism);
+		receiver.setInvokableClass(Receiver.class);
+
+		final SlotSharingGroup slotSharingGroup = new SlotSharingGroup();
+		receiver.setSlotSharingGroup(slotSharingGroup);
+		sender.setSlotSharingGroup(slotSharingGroup);
+
+		receiver.setStrictlyCoLocatedWith(sender);
```

Comment:
Even without L82-86 the test can pass. The reason is that some Sender/Receiver parallelism start and finish quickly. We can make sure that All Senders don't exit until all Receivers become running, maybe by setting a CountDownLatch like #6883

Reply:
this is true. We don't have true assertions making sure that task are being co-located. The `CountDownLatch` would enforce that both tasks are online at the same time. I think this is not what we want to guarantee here. Instead we should test that the tasks are deployed in the same slot and, thus, using local channels for communication. Maybe a non serializable record could do the trick here. I'll try it out.

Reply:
Hmm this doesn't work because we always serialize into a buffer independent of the channel type. The only difference is whether it goes through Netty or not I think.

Reply:
I think this should be the solution. What we can do is to start the `MiniCluster` with only local communication enabled. That way we won't start netty and the communication needs to happen strictly locally :-).

Reply:
make sense :-)

Reply:
@tillrohrmann The test still succeeds even if local communication is set to false.

Reply:
it's expected that the test succeeds if `localCommunication` is set to `false` because it's the less restricted case. If `localCommunication` is `true` TMs cannot speak with each other.

Reply:
of course, that makes more sense 🤦‍♂️ . Unfortunately the test still runs successfully if the colocation constraint is removed. Based on the logs the sender tasks are finishing before the receivers are even started, so we never run out of slots, which as I understand is the failure condition here.

Reply:
As @zentol and me talked offline, the test actually tests not only the co-location constraints but also the input preferences of normal scheduling. Thus, one needs to remove the slot sharing as well in order to make this test fail.